### PR TITLE
CellTenant targets ReplyAddress

### PR DESCRIPTION
**Pull request by Harwayne**
Helps with #2037.

## Proposed Changes

- CellTenant targets can specify their own `replyAddress`, rather than always using the CellTenant's address.
- Messages can be delivered to the `replyAddress` without needing the subscriber `address`.

**Release Note**

```release-note
NONE
```

### CODE REVIEW

**Comment by Harwayne** [comment]
```
Error from server (InternalError): error when creating "https://storage.googleapis.com/knative-nightly/net-istio/latest/net-istio.yaml": Internal error occurred: failed calling webhook "config.webhook.serving.knative.dev": Post https://webhook.knative-serving.svc:443/config-validation?timeout=10s: no endpoints available for service "webhook"
```

/retest

**Comment by Harwayne** [comment]
This is ready for review.

**Comment by Harwayne** [comment]
/retest

The failures all appear to be ksvc failures.

**Comment by Harwayne** [comment]
/retest

The failures all appear to be ksvc failures.

File: pkg/broker/handler/processors/deliver/processor.go [hunk 1]
```diff
@@ -147,23 +147,79 @@ func (p *Processor) Process(ctx context.Context, e *event.Event) error {
...
 // deliver delivers msg to target and sends the target's reply to the broker ingress.
 func (p *Processor) deliver(ctx context.Context, target *config.Target, broker *config.CellTenant, msg binding.Message, hops int32) error {
+	// Channels can have a reply address without a subscriber. So default the replyMessage to the
+	// original message. If there is a subscriber, then replyMessage is overwritten.
+	replyMessage := msg
+	if target.Address != "" {
+		replyMsg, cleanUp, err := p.sendToSubscriber(ctx, target, msg, hops)
+		defer cleanUp()
+		if err != nil {
+			return fmt.Errorf("sending event to subscriber: %w", err)
```
Comment (zhongduo) [inline pkg/broker/handler/processors/deliver/processor.go:157]: Nits: `failed to send event....`

File: pkg/broker/config/targets.proto [hunk 2]
```diff
@@ -87,6 +87,9 @@ message Target {
   // The resolved subscriber URI of the target.
   string address = 5;
 
+  // The resolved URI that replies are sent to.
+  string reply_address = 10;
```
[the remark this answers is not in the document]
Comment (Harwayne) [reply pkg/broker/config/targets.proto:91]: When you say `reorder` do you mean altering the numbers, such that reply_address is 6 rather than 10? Or moving the fields in this file to match the number order? Or something else?

[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/broker/config/targets.proto:91]: I was thinking about moving fields to match the number order, so that we don't have to find the max number manually next time we add one.

File: pkg/broker/handler/processors/deliver/processor.go [hunk 3]
```diff
@@ -147,23 +147,79 @@ func (p *Processor) Process(ctx context.Context, e *event.Event) error {
 
 // deliver delivers msg to target and sends the target's reply to the broker ingress.
 func (p *Processor) deliver(ctx context.Context, target *config.Target, broker *config.CellTenant, msg binding.Message, hops int32) error {
+	// Channels can have a reply address without a subscriber. So default the replyMessage to the
+	// original message. If there is a subscriber, then replyMessage is overwritten.
+	replyMessage := msg
+	if target.Address != "" {
```
[the remark this answers is not in the document]
Comment (Harwayne) [reply pkg/broker/handler/processors/deliver/processor.go:153]: Great catch. Added a special case and a unit test.

File: pkg/broker/handler/processors/deliver/processor.go (hunk 1 above)
Comment (Harwayne) [reply pkg/broker/handler/processors/deliver/processor.go]: Done.

File: pkg/broker/handler/processors/deliver/processor.go [hunk 4]
```diff
@@ -147,23 +147,83 @@ func (p *Processor) Process(ctx context.Context, e *event.Event) error {
...
+	var transformers []binding.Transformer
+	if target.CellTenantType == config.CellTenantType_BROKER {
+		// Hops only exist for the Broker. Nothing else uses them.
+		// Attach the previous hops for the reply.
+		transformers = append(transformers, eventutil.SetRemainingHopsTransformer(hops))
+	}
+
+	replyResp, err := p.sendMsg(ctx, replyAddress, replyMessage, transformers...)
+	if err != nil {
+		return fmt.Errorf("sending event to reply: %w", err)
```
Comment (zhongduo) [inline pkg/broker/handler/processors/deliver/processor.go:191]: failed here too?

File: pkg/broker/config/targets.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (Harwayne) [reply pkg/broker/config/targets.proto]: Done.

File: pkg/broker/handler/processors/deliver/processor.go (hunk 4 above)
Comment (Harwayne) [reply pkg/broker/handler/processors/deliver/processor.go]: Done.

File: pkg/broker/handler/processors/deliver/processor.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/broker/handler/processors/deliver/processor.go:153]: So after the latest commit, we will nack the message to the delivery pubsub in the broker topic and then reply on pubsub to redeliver it, right? The previous behavior is move to the DLQ instead. I think both way are good, but just want to make sure we understand the difference.

[the remark this answers is not in the document]
Comment (Harwayne) [reply pkg/broker/handler/processors/deliver/processor.go:153]: This behavior should be identical to before. In both cases, an empty string address results in `deliver` returning an error. In the past, this was because we actually tried sending the event and the HTTP call failed saying `""` is a bad address https://github.com/google/knative-gcp/blob/c19b557da0480a1806b07243bcbb76dd7ee12413/pkg/broker/handler/processors/deliver/processor.go#L152-L160
Now we just generate the error directly. 

The DLQ is handled by the PubSub layer, not the delivery processor. Both before and after this PR. When we create the Subscription, we instruct PubSub to send to a DLQ after N failed deliveries.

[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/broker/handler/processors/deliver/processor.go:153]: Sorry, my bad, I meant to say retry queue instead of DLQ. You are right both case it will be sent to the retry topic in lin 142.

File: pkg/broker/handler/processors/deliver/processor_test.go [hunk 5]
```diff
@@ -272,6 +272,63 @@ func (h *statusCodeReplyHandler) ServeHTTP(w http.ResponseWriter, req *http.Requ
 	}
 }
 
+func TestProcess_TriggerWithoutSubscriberAddress(t *testing.T) {
```
Comment (zhongduo) [inline pkg/broker/handler/processors/deliver/processor_test.go:275]: Should we also test the empty address for other CellTenant type (channel) to make sure that it will be delivered to the specified reply address? Also, what if the reply address is empty too?

Nits: we are not using "_" in test names.

**Comment by zhongduo** [comment]
/lgtm
/hold 
in case you want to add more test as suggested in this PR.

File: pkg/broker/handler/processors/deliver/processor_test.go (hunk 5 above)
Comment (Harwayne) [reply pkg/broker/handler/processors/deliver/processor_test.go]: We should have a test that shows that. Unfortunately, I can't at the moment. There are only two values for the CellTenantType enum, BROKER and UNKNOWN. We have defaulting logic that pretends UNKNOWNs are actually BROKER. So for now, I can't add this test.

However, I set up the test to make it possible and included a commented out version of what the test will be.

**Comment by Harwayne** [comment]
/unhold

Addressed the comment.
